src/main/java/com/ning/http/client/providers/grizzly/PauseContextHelper.java

```diff
@@ -0,0 +1,132 @@
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package com.ning.http.client.providers.grizzly;
+
+import org.glassfish.grizzly.filterchain.FilterChainContext;
+import org.glassfish.grizzly.filterchain.InvokeAction;
+import org.glassfish.grizzly.filterchain.NextAction;
+
+/**
+ * Package-private helper used to pause and resume the event processing of a {@link FilterChainContext}.
+ * Apart from the base suspend/resume mechanism from Grizzly, what this helper offers is
+ * {@link #pauseIfNeeded(FilterChainContext,NextAction) a method} to attach an {@link NextAction action} to the
+ * context, and {@link #resumeFromPausedAction(FilterChainContext) another method} to resume the execution starting
+ * from that attached action.
+ * This class is intended to be used only by {@link AhcEventFilter}.
+ */
+final class PauseContextHelper {
+
+    private PauseContextHelper() {
+        // private empty constructor to avoid wrong instantiation.
+    }
+
+    /**
+     * Marks the {@link FilterChainContext} so that the event filter knows that should pause the event processing after
+     * the current action.
+     * @param ctx the {@link FilterChainContext} to be paused.
+     */
+    public static void requestPause(FilterChainContext ctx) {
+        synchronized (ctx) {
+            PauseContext pauseContext = getPauseCtxFromAttribute(ctx);
+            if (null == pauseContext) {
+                setPauseCtxAttribute(ctx, new PauseContext(null));
+            } else {
+                throw new IllegalStateException("Can't pause an already paused context");
+            }
+        }
+    }
+
+    /**
+     * If the context has to be paused, it attaches the passed pausedAction to the {@link FilterChainContext} so that
+     * it can be resumed starting from that action later.
+     * @param ctx the {@link FilterChainContext} being paused.
+     * @param pausedAction the action to save.
+     * @return if the context has to be paused, it returns a suspend action. Otherwise, it just returns the same action
+     *         that was received as parameter.
+     */
+    public static NextAction pauseIfNeeded(FilterChainContext ctx, NextAction pausedAction) {
+        if (null == pausedAction || InvokeAction.TYPE != pausedAction.type()) {
+            return pausedAction;
+        }
+
+        if (null == getPauseCtxFromAttribute(ctx)) {
+            return pausedAction;
+        }
+
+        synchronized (ctx) {
+            PauseContext pauseContext = getPauseCtxFromAttribute(ctx);
+            if (pauseContext != null && pauseContext.getPausedAction() != null) {
+                throw new IllegalStateException("Can't override a paused action");
+            }
+
+            if (pauseContext == null) {
+                // it was resumed before this method could be executed
+                return pausedAction;
+            }
```

Comment:
I think if we swap the order of this ifs we can simplify the one about the override by not having to check the pauseContext being not null.
```suggestion
            if (pauseContext == null) {
                // it was resumed before this method could be executed
                return pausedAction;
            }
            
            if (pauseContext.getPausedAction() != null) {
                throw new IllegalStateException("Can't override a paused action");
            }

```